File: native/csrc/temporal_auth.c
/*
 * Temporal Auth Tree implementation for QRTB
 */

#include "temporal_auth.h"
#include "sha3.h"
#include <string.h>
#include <stdlib.h>

/* Derive key seed for index within batch */
static void derive_key_seed(const uint8_t batch_seed[64], size_t index,
                            uint8_t out[64]) {
    uint8_t buf[8 + 64 + 8]; /* "key_seed" + batch_seed + index_be64 */
    memcpy(buf, "key_seed", 8);
    memcpy(buf + 8, batch_seed, 64);
    buf[72] = (uint8_t)((index >> 56) & 0xFF);
    buf[73] = (uint8_t)((index >> 48) & 0xFF);
    buf[74] = (uint8_t)((index >> 40) & 0xFF);
    buf[75] = (uint8_t)((index >> 32) & 0xFF);
    buf[76] = (uint8_t)((index >> 24) & 0xFF);
    buf[77] = (uint8_t)((index >> 16) & 0xFF);
    buf[78] = (uint8_t)((index >> 8) & 0xFF);
    buf[79] = (uint8_t)(index & 0xFF);
    sha3_512(buf, 80, out);
}

void tat_derive_initial_batch_seed(const uint8_t *master_seed, size_t seed_len,
                                   uint8_t out[64]) {
    /* SHA3-512("batch_initial" || master_seed) */
    size_t buf_len = 13 + seed_len;
    uint8_t *buf = (uint8_t *)malloc(buf_len);
    memcpy(buf, "batch_initial", 13);
    memcpy(buf + 13, master_seed, seed_len);
    sha3_512(buf, buf_len, out);
    free(buf);
}

void tat_derive_next_batch_seed(const uint8_t current[64], uint8_t out[64]) {
    /* SHA3-512("batch_next" || current) */
    uint8_t buf[10 + 64];
    memcpy(buf, "batch_next", 10);
    memcpy(buf + 10, current, 64);
    sha3_512(buf, 74, out);
}

void tat_init(TemporalAuthTree *tree, const uint8_t batch_seed[64]) {
    memcpy(tree->batch_seed, batch_seed, 64);
    tree->key_index = 0;
    tree->rotation_index = 0;

    /* Allocate and generate all public keys */
    tree->pub_keys = (uint8_t *)malloc(TAT_TOTAL_KEYS * WOTS_SIG_SIZE);
    uint8_t private_key[WOTS_SIG_SIZE];

    for (int i = 0; i < TAT_TOTAL_KEYS; i++) {
        uint8_t key_seed[64];
        derive_key_seed(batch_seed, i, key_seed);
        wots_keygen(key_seed, private_key,
                    tree->pub_keys + i * WOTS_SIG_SIZE);
    }

    /* Pre-compute all Merkle proofs */
    tree->proofs = (MerkleProof *)malloc(TAT_TOTAL_KEYS * sizeof(MerkleProof));
    for (int i = 0; i < TAT_TOTAL_KEYS; i++) {
        uint8_t root_tmp[MERKLE_HASH_SIZE];
        merkle_build_with_proof(tree->pub_keys, TAT_TOTAL_KEYS, WOTS_SIG_SIZE,
                                i, root_tmp, &tree->proofs[i]);
        if (i == 0) {
            memcpy(tree->auth_root, root_tmp, MERKLE_HASH_SIZE);
        }
    }

    /* Clear private key from stack */
    memset(private_key, 0, WOTS_SIG_SIZE);
}

int tat_sign(TemporalAuthTree *tree, const uint8_t *message, size_t msg_len,
             uint8_t signature[WOTS_SIG_SIZE], uint8_t pub_key[WOTS_SIG_SIZE],
             MerkleProof *proof, size_t *key_index_out) {
    if (tree->key_index >= TAT_USABLE_KEYS)
        return -1;

    size_t idx = tree->key_index++;
    *key_index_out = idx;

    /* Derive key and sign */
    uint8_t key_seed[64];
    derive_key_seed(tree->batch_seed, idx, key_seed);
    uint8_t private_key[WOTS_SIG_SIZE];
    wots_keygen(key_seed, private_key, pub_key);
    wots_sign(message, msg_len, private_key, signature);

    /* Use pre-cached proof */
    memcpy(proof, &tree->proofs[idx], sizeof(MerkleProof));

    /* Clear sensitive data */
    memset(private_key, 0, WOTS_SIG_SIZE);
    memset(key_seed, 0, 64);

    return 0;
}

int tat_sign_rotation(TemporalAuthTree *tree, const uint8_t *message, size_t msg_len,
                      uint8_t signature[WOTS_SIG_SIZE], uint8_t pub_key[WOTS_SIG_SIZE],
                      MerkleProof *proof, size_t *key_index_out) {
    if (tree->rotation_index >= TAT_RESERVED)
        return -1;

    size_t idx = TAT_USABLE_KEYS + tree->rotation_index++;
    *key_index_out = idx;

    uint8_t key_seed[64];
    derive_key_seed(tree->batch_seed, idx, key_seed);
    uint8_t private_key[WOTS_SIG_SIZE];
    wots_keygen(key_seed, private_key, pub_key);
    wots_sign(message, msg_len, private_key, signature);

    /* Use pre-cached proof */
    memcpy(proof, &tree->proofs[idx], sizeof(MerkleProof));

    memset(private_key, 0, WOTS_SIG_SIZE);
    memset(key_seed, 0, 64);

    return 0;
}

int tat_verify(const uint8_t *message, size_t msg_len,
               const uint8_t signature[WOTS_SIG_SIZE],
               const uint8_t pub_key[WOTS_SIG_SIZE],
               const MerkleProof *proof,
               const uint8_t auth_root[MERKLE_HASH_SIZE]) {
    /* Verify WOTS+ signature */
    if (!wots_verify(message, msg_len, signature, pub_key))
        return 0;

    /* Verify Merkle proof: hash the pub_key to get leaf hash */
    uint8_t leaf_hash[MERKLE_HASH_SIZE];
    sha3_256(pub_key, WOTS_SIG_SIZE, leaf_hash);

    return merkle_verify_proof(leaf_hash, proof, auth_root);
}

void tat_destroy(TemporalAuthTree *tree) {
    memset(tree->batch_seed, 0, 64);
    if (tree->pub_keys) {
        free(tree->pub_keys);
        tree->pub_keys = NULL;
    }
    if (tree->proofs) {
        free(tree->proofs);
        tree->proofs = NULL;
    }
    tree->key_index = 0;
    tree->rotation_index = 0;
}

File: native/src/temporal_auth.rs
use crate::sha3::{sha3_256, sha3_512, HASH_SIZE_512};
use crate::wots::{wots_keygen, wots_sign, wots_verify};
use crate::merkle::{merkle_build_with_proof, merkle_verify_proof, MerkleProof, MERKLE_HASH_SIZE};
use zeroize::Zeroize;

pub const TOTAL_KEYS_PER_BATCH: usize = 1024;
pub const RESERVED_ROTATION: usize = 2;
pub const USABLE_KEYS: usize = TOTAL_KEYS_PER_BATCH - RESERVED_ROTATION; // 1022

/// Derive initial batch seed from master seed
pub fn derive_initial_batch_seed(master_seed: &[u8]) -> [u8; HASH_SIZE_512] {
    let mut buf = Vec::with_capacity(14 + master_seed.len());
    buf.extend_from_slice(b"batch_initial");
    buf.extend_from_slice(master_seed);
    sha3_512(&buf)
}

/// Derive next batch seed (forward secrecy chain)
pub fn derive_next_batch_seed(current_batch_seed: &[u8; HASH_SIZE_512]) -> [u8; HASH_SIZE_512] {
    let mut buf = Vec::with_capacity(10 + 64);
    buf.extend_from_slice(b"batch_next");
    buf.extend_from_slice(current_batch_seed);
    sha3_512(&buf)
}

/// Derive key seed for a specific index within a batch
fn derive_key_seed(batch_seed: &[u8; HASH_SIZE_512], index: usize) -> [u8; HASH_SIZE_512] {
    let mut buf = Vec::with_capacity(8 + 64 + 8);
    buf.extend_from_slice(b"key_seed");
    buf.extend_from_slice(batch_seed);
    buf.extend_from_slice(&(index as u64).to_be_bytes());
    sha3_512(&buf)
}

/// Temporal authentication signature result
pub struct TemporalSignature {
    pub signature: Vec<u8>,
    pub public_key: Vec<u8>,
    pub proof: MerkleProof,
    pub key_index: usize,
}

/// Temporal Auth Tree: manages a batch of 1024 WOTS+ keys
pub struct TemporalAuthTree {
    batch_seed: [u8; HASH_SIZE_512],
    auth_root: [u8; MERKLE_HASH_SIZE],
    key_index: usize,
    rotation_index: usize,
    /// Cached public keys
    pub_keys: Vec<Vec<u8>>,
    /// Pre-computed Merkle proofs for all keys
    proofs: Vec<MerkleProof>,
}

impl TemporalAuthTree {
    /// Create a new TemporalAuthTree from a batch seed
    pub fn new(batch_seed: [u8; HASH_SIZE_512]) -> Self {
        // Generate all public keys
        let mut pub_keys = Vec::with_capacity(TOTAL_KEYS_PER_BATCH);
        for i in 0..TOTAL_KEYS_PER_BATCH {
            let key_seed = derive_key_seed(&batch_seed, i);
            let (_, pk) = wots_keygen(&key_seed);
            pub_keys.push(pk);
        }

        // Build all proofs at once
        let leaves: Vec<&[u8]> = pub_keys.iter().map(|k| k.as_slice()).collect();
        let mut proofs = Vec::with_capacity(TOTAL_KEYS_PER_BATCH);
        let mut root = [0u8; MERKLE_HASH_SIZE];
        for i in 0..TOTAL_KEYS_PER_BATCH {
            let (r, proof) = merkle_build_with_proof(&leaves, i);
            if i == 0 { root = r; }
            proofs.push(proof);
        }

        TemporalAuthTree {
            batch_seed,
            auth_root: root,
            key_index: 0,
            rotation_index: 0,
            pub_keys,
            proofs,
        }
    }

    /// Get the auth root (Merkle root of all public keys)
    pub fn auth_root(&self) -> &[u8; MERKLE_HASH_SIZE] {
        &self.auth_root
    }

    /// Get remaining usable keys
    pub fn remaining_keys(&self) -> usize {
        USABLE_KEYS - self.key_index
    }

    /// Sign a message using the next available key (keys 0..1021)
    pub fn sign(&mut self, message: &[u8]) -> Option<TemporalSignature> {
        if self.key_index >= USABLE_KEYS {
            return None; // Exhausted
        }

        let idx = self.key_index;
        self.key_index += 1;

        let key_seed = derive_key_seed(&self.batch_seed, idx);
        let (sk, pk) = wots_keygen(&key_seed);
        let sig = wots_sign(message, &sk);

        Some(TemporalSignature {
            signature: sig,
            public_key: pk,
            proof: self.proofs[idx].clone(),
            key_index: idx,
        })
    }

    /// Sign with a reserved rotation key (keys 1022..1023)
    pub fn sign_rotation(&mut self, message: &[u8]) -> Option<TemporalSignature> {
        if self.rotation_index >= RESERVED_ROTATION {
            return None;
        }

        let idx = USABLE_KEYS + self.rotation_index;
        self.rotation_index += 1;

        let key_seed = derive_key_seed(&self.batch_seed, idx);
        let (sk, pk) = wots_keygen(&key_seed);
        let sig = wots_sign(message, &sk);

        Some(TemporalSignature {
            signature: sig,
            public_key: pk,
            proof: self.proofs[idx].clone(),
            key_index: idx,
        })
    }

    /// Verify a temporal auth signature against this tree's root
    pub fn verify(&self, message: &[u8], sig: &TemporalSignature) -> bool {
        Self::verify_against_root(message, sig, &self.auth_root)
    }

    /// Verify against an arbitrary auth root
    pub fn verify_against_root(
        message: &[u8],
        sig: &TemporalSignature,
        auth_root: &[u8; MERKLE_HASH_SIZE],
    ) -> bool {
        // Verify WOTS+ signature
        if !wots_verify(message, &sig.signature, &sig.public_key) {
            return false;
        }

        // Verify Merkle proof
        let leaf_hash = sha3_256(&sig.public_key);
        merkle_verify_proof(&leaf_hash, &sig.proof, auth_root)
    }

    /// Destroy the batch seed (forward secrecy)
    pub fn destroy(&mut self) {
        self.batch_seed.zeroize();
        self.pub_keys.clear();
        self.proofs.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_batch_seed_derivation() {
        let master = sha3_512(b"master_seed");
        let batch0 = derive_initial_batch_seed(&master);
        let batch1 = derive_next_batch_seed(&batch0);
        let batch2 = derive_next_batch_seed(&batch1);

        // All different
        assert_ne!(batch0, batch1);
        assert_ne!(batch1, batch2);

        // Deterministic
        let batch0_again = derive_initial_batch_seed(&master);
        assert_eq!(batch0, batch0_again);
    }

    #[test]
    fn test_sign_verify() {
        let seed = derive_initial_batch_seed(b"test_temporal");
        let mut tree = TemporalAuthTree::new(seed);

        let msg = b"test message";
        let sig = tree.sign(msg).unwrap();
        assert!(tree.verify(msg, &sig));
    }

    #[test]
    fn test_wrong_message_fails() {
        let seed = derive_initial_batch_seed(b"test_temporal_2");
        let mut tree = TemporalAuthTree::new(seed);

        let sig = tree.sign(b"message A").unwrap();
        assert!(!tree.verify(b"message B", &sig));
    }

    #[test]
    fn test_key_exhaustion() {
        let seed = derive_initial_batch_seed(b"exhaust_test");
        let mut tree = TemporalAuthTree::new(seed);

        // Use a few keys
        for _ in 0..5 {
            assert!(tree.sign(b"msg").is_some());
        }
        assert_eq!(tree.remaining_keys(), USABLE_KEYS - 5);
    }

    #[test]
    fn test_rotation_keys() {
        let seed = derive_initial_batch_seed(b"rotation_test");
        let mut tree = TemporalAuthTree::new(seed);

        let rot1 = tree.sign_rotation(b"rotate1").unwrap();
        assert!(tree.verify(b"rotate1", &rot1));

        let rot2 = tree.sign_rotation(b"rotate2").unwrap();
        assert!(tree.verify(b"rotate2", &rot2));

        // Third rotation should fail
        assert!(tree.sign_rotation(b"rotate3").is_none());
    }

    #[test]
    fn test_destroy_zeroes_seed() {
        let seed = derive_initial_batch_seed(b"destroy_test");
        let mut tree = TemporalAuthTree::new(seed);
        tree.destroy();
        assert_eq!(tree.batch_seed, [0u8; 64]);
        assert!(tree.pub_keys.is_empty());
    }

    #[test]
    fn test_verify_against_root() {
        let seed = derive_initial_batch_seed(b"root_verify");
        let mut tree = TemporalAuthTree::new(seed);
        let root = *tree.auth_root();

        let sig = tree.sign(b"check_root").unwrap();
        assert!(TemporalAuthTree::verify_against_root(b"check_root", &sig, &root));
    }
}
